dali/operators/image/crop/crop_attr.cc

```diff
@@ -70,4 +69,159 @@ for ``crop_w``, ``crop_h``, and ``crop_d`` is incompatible with providing the fi
 window dimensions (argument `crop`).)code",
         0.0f, true);
 
+CropAttr::CropAttr(const OpSpec& spec) {
+  auto max_batch_size = spec.GetArgument<int>("max_batch_size");
+  bool has_crop_arg = spec.ArgumentDefined("crop");
+  bool has_crop_w_arg = spec.ArgumentDefined("crop_w");
+  bool has_crop_h_arg = spec.ArgumentDefined("crop_h");
+  bool has_crop_d_arg = spec.ArgumentDefined("crop_d");
+  is_whole_image_ = !has_crop_arg && !has_crop_w_arg && !has_crop_h_arg && !has_crop_d_arg;
+
+  DALI_ENFORCE(has_crop_w_arg == has_crop_h_arg,
+               "`crop_w` and `crop_h` arguments must be provided together");
+
+  if (has_crop_d_arg) {
+    DALI_ENFORCE(has_crop_w_arg,
+                 "`crop_d` argument must be provided together with `crop_w` and `crop_h`");
+  }
+
+  int crop_d = kNoCrop, crop_h = kNoCrop, crop_w = kNoCrop;
+  if (has_crop_arg) {
+    DALI_ENFORCE(!has_crop_h_arg && !has_crop_w_arg && !has_crop_d_arg,
+                 "`crop` argument is not compatible with `crop_h`, `crop_w`, `crop_d`");
+    if (spec.HasArgument("crop")) {
+      auto crop = spec.GetRepeatedArgument<float>("crop");
+      DALI_ENFORCE(crop.size() >= 2 && crop.size() <= 3,
+                   "`crop` argument should have 2 or 3 elements depending on the input data shape");
+      int i = 0;
+      has_crop_d_ = crop.size() == 3;
+      if (has_crop_d_) {
+        crop_d = crop[i++];
+      }
+      crop_h = crop[i++];
+      crop_w = crop[i++];
+    }
+  } else {
+    has_crop_d_ = has_crop_d_arg;
+  }
+  crop_height_.resize(max_batch_size, crop_h);
+  crop_width_.resize(max_batch_size, crop_w);
+  crop_depth_.resize(max_batch_size, crop_d);
+  crop_x_norm_.resize(max_batch_size, 0.0f);
+  crop_y_norm_.resize(max_batch_size, 0.0f);
+  crop_z_norm_.resize(max_batch_size, 0.0f);
+  crop_window_generators_.resize(max_batch_size, {});
+}
+
+void CropAttr::ProcessArguments(const OpSpec& spec, const ArgumentWorkspace* ws,
+                                std::size_t data_idx) {
+  int crop_arg_len = 0;
+  if (spec.HasTensorArgument("crop")) {
+    auto crop_arg = view<const float, 1>(ws->ArgumentInput("crop"))[data_idx];
+    crop_arg_len = crop_arg.shape[0];
+    DALI_ENFORCE(crop_arg_len >= 2 && crop_arg_len <= 3,
+                 "`crop` argument should have 2 or 3 elements depending on the input data shape");
+    int idx = 0;
+    has_crop_d_ = crop_arg_len == 3;
```

Comment:
Just a loose thought that at a first glance it seems slightly convoluted now. For instance where `has_crop_d_` comes from in particular tensor/spec case.

Reply:
You are right. I'll clean up